tracks/julia/exercises/word-count/mentoring.md

```diff
@@ -121,4 +121,94 @@ function wordcount(sentence)
 end
 ```
 
+** Without Regex ** 
+
+This solution doesn't use a regex at all. It uses the default split behavior (which checks for `isspace` characters) and then strips punctuation characters left over on the "outside" of words. Filtering hoas to come after stripping in case the strip operation itself left an empty string.
+```julia
+function wordcount(sentence)
+    words = strip.(ispunct, split(lowercase(sentence)))
+    filter!(!isempty, words)
+    d = Dict{eltype(words), Int}()
+    for w in words
+        d[w] = get(d, w, 0) + 1
+    end
+    d
+end
+```
+It's possible to create the dictionary more efficiently with the `countmap` function from `StatsBase` (which only hashes the key once per check). With countmap, you can code-golf the above into:
+```julia
+wordcount_withstatsbase(sentence) = countmap(filter!(!isempty, strip.(ispunct, split(lowercase(sentence)))))
+```
+
+
+** Totally manual ** 
+
+This method iterates through the string one character at a time and switches behavior at word boundaries (when `iswordchar` changes from true to false or false to true). The dictionary has `SubString` keys in this version, which are like the string-version of Array `views`.
+```julia
+function wordcount(sent)
+
+    sent = lowercase(sent)
+    d = Dict{SubString, Int}()
+
+    in_word = false; j = 0
+    # For our purposes, almost everything is a valid word character.
+    iswordchar(c) = !ispunct(c) && !isspace(c)
+    for i in eachindex(sent)
+        c = sent[i]
+        if iswordchar(c) && i < lastindex(sent)
+            # if we're not already in a word, then this character marks
+            # the start of a new word. If we are in a word, just keep going.
+            if !in_word
+                in_word = true
+                j = i
+            end
+        elseif in_word
```

Comment:
continue and if might be a little clearer than elseif?

I had to look twice seeing in_word set and then tested.

Reply:
I'm assuming this is what you meant. I'm not sure it's necessarily more readable, since the second `if` only happens if the first doesn't, and the `continue` may be hard to miss. I'll give it a minute and then look at it again.

```julia
function wordcount6_4(input)

    input = lowercase(input)
    d = Dict{SubString{String}, Int}()

    in_word = false; j = 0
    # For our purposes, almost everything is a valid word character.
    iswordchar(c) = !ispunct(c) && !isspace(c)
    for i in eachindex(input)
        c = input[i]
        
        if iswordchar(c)
            # if we're not already in a word, then this character marks
            # the start of a new word. If we are in a word, just keep going.
            if i < lastindex(input) && !in_word
                in_word = true
                j = i
            end
            continue
        end

    
        if in_word
            # If `c` is not a word character, but we are in a word, then the
            # previous character marks the end of a word. However, if the current
            # character is a ' we need to see if this word is a contraction by
            # looking at the next character
            if c == '''
                i < lastindex(input) &&
                iswordchar(input[nextind(input, i)]) &&
                continue
            end

            in_word = false
            k = SubString(input, j:(i-1))
            d[k] = get(d, k, 0) + 1
        end
    end
    d
end
```

Reply:
I think I prefer this version, but please choose whichever you prefer, this was just a suggestion :)